### PR TITLE
Add hook to send logs to custom writer #678

### PR DESCRIPTION
**Pull request by bunyk**
Implements #678

### CODE REVIEW

**Comment by gunnihinn** [comment]
LGTM (fwiw).

**Comment by bunyk** [comment]
@tgwizard

One way of solving overlapping log levels would be replacing `Writer` field with the slice of `Writers`, and after converting to bytes send to each in loop. So then you would be able for example configure three hooks: one for sdtout, one for stderr, and one for both.

**Comment by bunyk** [comment]
@tgwizard  handling overlapping log levels  by allowing slice of writers in hook adds another challenge - error handling. Should we try to write to each writer, and then return `Error` struct that wraps any errors that happen, like https://github.com/hashicorp/go-multierror ?

**Comment by elliotaplant** [comment]
Love this PR, but it looks like the appveyor build got cancelled. Is there anything left besides re-running appveyor left before this can get merged?

File: hooks/writer/writer.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// Hook is a hook that writes logs of specified LogLevels to specified Writer
+type Hook struct {
+	Writer    io.Writer
+	LogLevels []log.Level
+}
+
+// Fire will be called when some logging function is called with current hook
+// It will format log entry to string and write it to appropriate writer
+func (hook *Hook) Fire(entry *log.Entry) error {
+	line, err := entry.Bytes()
```
Comment (alphaXkiller) [inline hooks/writer/writer.go]: Can we just use `entry.Buffer.Bytes()`?
```go
  _, err = hook.Writer.Write(entry.Buffer.Bytes()
  return err
```

**Comment by ashleyschuett** [comment]
+1 what's the plan for this PR?

**Comment by bunyk** [comment]
> +1 what's the plan for this PR?

I don't know how to restart  AppVeyor build. Commit some unrelevant change?

**Comment by eberkund** [comment]
This looks awesome! I have been using `io.MultiWriter` but that has the unwanted side effect of wrecking the nice terminal colors for `os.Stdout` output.

**Comment by eberkund** [comment]
> Commit some unrelevant change?

That would do it, you can always rebase before merging anyhow.

**Comment by bt** [comment]
Any chance this will get merged anytime soon? Want to roll it out into my production app for different logging level files.

**Comment by bunyk** [comment]
> > Commit some unrelevant change?
> 
> That would do it, you can always rebase before merging anyhow.

Yep, that helped, all green. ;)

**Comment by bunyk** [comment]
@dgsb @lynncyrin What are the next steps? Review from you?

**Comment by lynncyrin** [comment]
@bunyk I'm not a maintainer here, so you'll want a review from `@dgsb`

**Comment by Pokerkoffer** [comment]
When will this be merged? Is there anything missing?

**Comment by bunyk** [comment]
@dgsb I guess I still need your review. :)

**Comment by bunyk** [comment]
@markphelps Tests are passing, reviewers approving. Who on this project presses decides to do merge?

**Comment by lavoiesl** [comment]
🚢

**Comment by markphelps** [comment]
Sorry for the stalebot spam. I thought it would only label/comment on issues. I updated the config to not label/comment on PRs in https://github.com/sirupsen/logrus/commit/494ec951d122f078df18d32aa520ba9b3f39b39f

**Comment by xuchen81** [comment]
@bunyk thanks for the solution, I am using exactly the same code here https://github.com/sirupsen/logrus/tree/master/hooks/writer, but it does not work for DebugLevel, it is not written in neither Studout nor Studerr.

**Comment by 27149chen** [comment]
> @bunyk thanks for the solution, I am using exactly the same code here https://github.com/sirupsen/logrus/tree/master/hooks/writer, but it does not work for DebugLevel, it is not written in neither Studout nor Studerr.

@xuchen81 , you should enable debug log by `log.SetLevel()` first

**Pull request merged**